Api instance methods (#65)
1950Labs/vuejs.org: src/v2/api/index.md

  - `{string | Function} expOrFn`
  - `{Function | Object} callback`
  - `{Object} [options]`
    - `{boolean} deep`
    - `{boolean} immediate`

- **Retorna:** `{Function} unwatch`

- **Uso:**

  Observa por cambios en una expresión o función computada en la instancia de Vue. La función _callback_ es llamada con el nuevo y el antiguo valor. La expresión solamente acepta caminos delimitados por puntos. Para expresiones más complejas, utilice una función.

<p class="tip">Nota: cuando se está alterando (en vez de reemplazando) un Objeto o un Arreglo, el antiguo valor será igual al nuevo, ya que estos referencian al mismo Objeto/Arreglo. Vue no mantiene una copia del valores antes de que estos sean mutados.</p>

- **Ejemplo:**

  ``` js
  // camino de la clave
  vm.$watch('a.b.c', function (newVal, oldVal) {
    // hacer algo
  })

  // función
  vm.$watch(
    function () {
      return this.a + this.b
    },
    function (newVal, oldVal) {
      // hacer algo
    }
  )
  ```

  `vm.$watch` retorna una función _unwatch_ que para de lanzar la función _callback_:

  ``` js
  var unwatch = vm.$watch('a', cb)
  // luego, destruir el `watcher`
  unwatch()
  ```

- **Opción: deep**

>   Para también detectar valores anidados dentro de Objetos, usted debe pasar `deep: true` en las opcoines. Note que usted no necesita realizar esto para mutaciones de Arreglos.

Reviewer: Minor typo here:
```suggestion
  Para también detectar valores anidados dentro de Objetos, usted debe pasar `deep: true` en las opciones. Note que usted no necesita realizar esto para mutaciones de Arreglos.
```

  ``` js
  vm.$watch('someObject', callback, {
    deep: true
  })
  vm.someObject.nestedValue = 123

>   // se dispará la función `callback`

Reviewer: I think you meant:

```suggestion
  // se disparará la función `callback`
```

  ```

- **Opción: immediate**

  Pasar `immediate: true` en las opciones disparará la función _callback_ inmediatamente con el nuevo valor de la expresión:

  ``` js
  vm.$watch('a', callback, {
    immediate: true
  })

>   // se dispará el `callback` inmediatamente con el nuevo valor de `a`

Reviewer: Same misspelling here:

```suggestion
  // se disparará el `callback` inmediatamente con el nuevo valor de `a`
```

### vm.$once( event, callback )

- **Argumentos:**
  - `{string} event`
  - `{Function} callback`

- **Uso:**

>   Escucho un evento personalizado, pero solamente una vez. La escucha será removida una vez que es disparada por primera vez.

Reviewer: I think it's better to change this for:

```suggestion
  Escucha por un evento personalizado, pero solamente una vez. El _listener_ será removido una vez que es disparado por primera vez.
```

What do you think?

### vm.$off( [event, callback] )

- **Argumentos:**
  - `{string | Array<string>} event` (arreglo solamente es soportado en 2.2.2+)
  - `{Function} [callback]`

- **Uso:**

>   Remover escuchas de evento(s) personalizado(s).

Reviewer: I agree with you it's difficult to translate this in a way it sounds natural. Maybe what we can do is translate it as "escuchas" but add `(_listener_)` next to it? 🤔

Author: I like that idea! @FLarra 😄

>   - Si ningún argumento es provisto, remueve todas las escuchas a eventos.

Reviewer: Maybe it's better:

```suggestion
  - Si ningún argumento es proporcionado, remueve todas las escuchas a eventos.
```

what do you think?

  </script>
  {% endraw %}

## Métodos de Instancia / Ciclo de vida

### vm.$mount( [elementOrSelector] )

- **Argumentos:**
  - `{Element | string} [elementOrSelector]`
  - `{boolean} [hydrating]`

- **Retorna:** `vm` - la propia instancia

- **Uso:**

  Si la instancia de Vue no recibió la opción `el` durante la instanciación, ella estará en el estado "no montada" sin un elemento del DOM asociado. `vm.$mount()` puede ser utilizado para comenzar a montar una instancia de Vue no montada.

>   Si el argumento `elementOrSelector` no es provisto, la plantilla será renderizara como un elemento "fuera del documento", y ustede deberá utilizar la API nativa del DOM para insertarla en el documento.

Reviewer: Typo:

```suggestion
  Si el argumento `elementOrSelector` no es provisto, la plantilla será renderizará como un elemento "fuera del documento", y usted deberá utilizar la API nativa del DOM para insertarla en el documento.
```

Reviewer: @Tintef you missed this typo + misspelling 😃

>   El método retorna la propia instancia para que usted pueda encadenar otros métodos de instacia luego.

Reviewer: Typo here:

```suggestion
  El método retorna la propia instancia para que usted pueda encadenar otros métodos de instancia luego.
```

- **Ejemplo:**

  ``` js
  var MyComponent = Vue.extend({
    template: '<div>Hello!</div>'
  })

  // crea y monta en #app (reemplazará #app)
  new MyComponent().$mount('#app')

  // el ejemplo de arriba es equivalente a:
  new MyComponent({ el: '#app' })

  // o, renderizar "fuera del documento" y agregarla luego:
  var component = new MyComponent().$mount()
  document.getElementById('app').appendChild(component.$el)
  ```

- **Vea también:**
  - [Diagrama de Ciclo de Vida](../guide/instance.html#Diagrama-del-Ciclo-de-vida)
  - [Renderizado del lado de Servidor](../guide/ssr.html)

### vm.$forceUpdate()

- **Uso:**

>   Fuerza a la instancia de Vue a renderizarse nuevamente. Note que esto no afecta a todos sus componentes hijo, solamente a la propia instancia y componentes hijos con contenido insertado en _slot_.

Reviewer: Maybe "hijo" should be in plural here?

```suggestion
  Fuerza a la instancia de Vue a renderizarse nuevamente. Note que esto no afecta a todos sus componentes hijos, solamente a la propia instancia y componentes hijos con contenido insertado en _slot_.
```

### vm.$nextTick( [callback] )

- **Argumentos:**
  - `{Function} [callback]`

- **Uso:**

>   Pospone la ejecución de la función _callback_ para luego del próximo ciclo de actualización del DOM. Utilize este método inmediatamente luego de modificar datos para esperar por la actualización del DOM. Este método es el mismo que el método global `Vue.nextTick`, excpeto que el contexto `this` de la función _callback_ es automáticamente vinculado a la instancia que invoco este método.

Reviewer: There is a misspelling here:

```suggestion
  Pospone la ejecución de la función _callback_ para luego del próximo ciclo de actualización del DOM. Utilice este método inmediatamente luego de modificar datos para esperar por la actualización del DOM. Este método es el mismo que el método global `Vue.nextTick`, excepto que el contexto `this` de la función _callback_ es automáticamente vinculado a la instancia que invoco este método.
```

Author: Yeah, I always that mistake 🤦‍♂ I know it's wrong, but I still manage to spell it wrong 😅

>   > Nuevo en 2.1.0+: retorna una promesa (_Promise_) si no fue provista una función _callback_ y estas son soportadas en el entorno de ejecución. Note que Vue no trae un _polyfill_ para promesas, por lo tanto, si su navegador no soporta promseas de forma nativa (mirándote, IE), usted deberá agregar un _polyfill_ por su cuenta.

Reviewer: There is a typo here:

```suggestion
  > Nuevo en 2.1.0+: retorna una promesa (_Promise_) si no fue provista una función _callback_ y estas son soportadas en el entorno de ejecución. Note que Vue no trae un _polyfill_ para promesas, por lo tanto, si su navegador no soporta promesas de forma nativa (mirándote, IE), usted deberá agregar un _polyfill_ por su cuenta.
```

- **Ejemplo:**

  ``` js
  new Vue({
    // ...
    methods: {
      // ...
      example: function () {
        // modificar datos
        this.message = 'changed'
        // el DOM no está actualizado todavía
        this.$nextTick(function () {

>           // el DOM está actualizaod

Reviewer: Typo :) 

```suggestion
          // el DOM está actualizado
```

          // `this` está vinculado a la instancia actual
          this.doSomethingElse()
        })
      }
    }
  })
  ```

- **Vea también:**
  - [Vue.nextTick](#Vue-nextTick)
  - [Cola de actualización asíncrona](../guide/reactivity.html#Cola-de-actualizacion-asincrona)

### vm.$destroy()

- **Uso:**

>   Destruir completamente una `vm`. Limpia todas sus conecciones con otras `vms` existentes, desvincula todas sus directivas y apaga todas sus escuchas a eventos.

Reviewer: Misspelling here:

```suggestion
  Destruir completamente una `vm`. Limpia todas sus conexiones con otras `vms` existentes, desvincula todas sus directivas y apaga todas sus escuchas a eventos.
```

Author: Damn 😅

>   Dispará los `hooks` `beforeDestroy` y `destroyed`.

Reviewer: It's:

```suggestion
  Dispara los `hooks` `beforeDestroy` y `destroyed`.
```
